### PR TITLE
New `Filesystem` namespace

### PR DESCRIPTION
**Pull request by distantnative**
## Status: ready for review 👀

New `Kirby\File` namespace featuring a new classes setup for handling file assets in Kirby:

#### `Kirby\Toolkit`
- `Kirby\Toolkit\File` takes care of everything CMS-unrelated. Gains a few more methods (e.g. `download`, `headers`) that couple it a bit more with other packages (e.g. `Kirby\Http`)

#### `Kirby\Filesystem`
- New namespace bundling several file-relates classes
- New class `Kirby\Filesystem\File` extends `Kirby\Toolkit\File` with some Cms-related functions (e.g. `::modified()` with handler, `::isResizable()`) - takes over a lot of functionality that was previously in `Kirby\Image\Image` even for non-images

- New trait `Kirby\Filesystem\IsFile` adds method-proxying to an asset object - either to `Kirby\Image\Image` or `Kirby\Filesystem\File` depending on the file type of the `$root` property
- `Kirby\Cms\FileFoundation` is removed - it gets replaced by `Kirby\Filesystem\IsFile` and all proxied methods of `Kirby\Filesystem\File`/`Kirby\Image\Image` which reach down to  `Kirby\Toolkit\File` as well

- `Kirby\Cms\Asset` has been moved to the new namespace as `Kirby\Filesystem\Asset`
- `Kirby\Cms\Filename` has been moved to the new namespace as `Kirby\Filesystem\Filename`

#### `Kirby\Image`
-  `Kirby\Image\Image` now extends `Kirby\Filesystem\File` and adds image-related methods (e.g. `dimensions`, `exif`)

## Breaking changes
- There are aliases for mapping old class names with new ones
- `Kirby\Image\Image::__toString()` returns an `` tag instead the file path.
- File hooks: the `$upload` parameter is a `Kirby\Filesystem\File` object, before `Kirby\Image\Image`
- `Kirby\Toolkit\File::header()` now returns `void` instead of an empty `string` by default
- The `Asset::alt()` method was removed as it didn't have a function on its own

## Related issues
- Closes https://github.com/getkirby/kirby/issues/3233

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Added in-code documentation (if needed)
- [x] CI passes (runs automatically when the PR is created or run `composer ci` locally)
  Running locally requires PHPUnit, PHP-CS-Fixer, Psalm, PHPCPD and PHPMD.

- [x] Checked whether the PR needs documentation, if needed added to the [release docs checklist](https://github.com/getkirby/getkirby.com/pulls)

## When merging
- [ ] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)
- [x] Add breaking changes to [release notes draft](https://github.com/getkirby/kirby/releases)

### CODE REVIEW

**Comment by bastianallgeier** [comment]
Ok, before we move on with this, how about introducing a Kirby\Filesystem namespace instead and we move Dir and F there as well?

**Comment by distantnative** [comment]
@bastianallgeier I wouldn't move Toolkit classes into other namespaces... I feel like Toolkit classes are somehow special and (mostly) universal - not bound to the Cms? But maybe that thinking is false.

**Comment by bastianallgeier** [comment]
To be honest, the toolkit is more like an "I don't know where to put it" namespace 🤣  Otherwise some HTTP stuff and other classes could also be in the toolkit.

**Comment by distantnative** [comment]
True. Ok, fine with that. I like `Filesystem` as it makes the distinction to the model clearer.

Aaaaand... breaking up Toolkit: so many more possibilities now! :D

**Comment by bastianallgeier** [comment]
Hahaha, keep in mind that we don't need to achieve all of this in 3.6

**Comment by distantnative** [comment]
We don't need to, but...

I have changed the name to `Filesystem`, I would leave moving `F`, `Dir` etc. for a separate PR.

File: config/components.php
Comment (lukasbestle) [inline config/components.php]:
```suggestion
     * @param string $src Root of the original file
     * @param string $dst Template string for the root to the desired destination
```

I also feel like the param should be named `$destination`, not `$dst` (it's not really Kirby-like to abbreviate param names).

File: src/Cms/File.php
Comment (lukasbestle) [inline src/Cms/File.php]: Why these changes? In our PHP code we use lowercase line comments throughout.

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]:
```suggestion
 * A representation of a file in the filesystem.
```

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]:
```suggestion
     * @param array|string|null $props Properties or deprecated `$root` string
     * @param string|null $url Deprecated argument, use `$props['url']` instead
```

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]:
```suggestion

```

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]: I think this method could be part of `Toolkit\File` as it's not CMS-specific.

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]: This part should be in the `Image` class IMO.

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]:
```suggestion
     * Returns the file's last modification time
```

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]:
```suggestion
     * Converts the object to an array
```

File: src/Filesystem/File.php
Comment (lukasbestle) [inline src/Filesystem/File.php]:
```suggestion
     * Returns the URL for the file object
```

File: src/Filesystem/IsFile.php
Comment (lukasbestle) [inline src/Filesystem/IsFile.php]:
```suggestion

```

File: src/Filesystem/IsFile.php
Comment (lukasbestle) [inline src/Filesystem/IsFile.php]:
```suggestion
        // public property access
```

File: src/Filesystem/IsFile.php
Comment (lukasbestle) [inline src/Filesystem/IsFile.php]:
```suggestion
        // asset method proxy
```

File: src/Filesystem/Name.php
Comment (lukasbestle) [inline src/Filesystem/Name.php]: If we name the package `Filesystem` (👍), I think the `Name` class should still be named `Filename`, otherwise the mental connection is lost. With `File\Name` it made sense, but `Filesystem\Name` is weird: What kind of name?

File: src/Image/Image.php
Comment (lukasbestle) [inline src/Image/Image.php]:
```suggestion
     * Converts the object to an array
```

File: src/Image/Image.php
Comment (lukasbestle) [inline src/Image/Image.php]:
```suggestion
     * Returns the URL for the file object
```

File: src/Toolkit/File.php
Comment (lukasbestle) [inline src/Toolkit/File.php]:
```suggestion
     * @param string|null $filename Optional filename for the download
```

File: src/Toolkit/File.php
Comment (lukasbestle) [inline src/Toolkit/File.php]: I feel like we should remove the `string` return value and return `null` if `$send` is `true`. The returned string is always empty because the body is set to an empty string the line above.

This is a breaking change, but IMO there's no point returning an empty string here. It's only confusing.

File: tests/Cms/Files/FileVersionTest.php
Comment (lukasbestle) [inline tests/Cms/Files/FileVersionTest.php]:
```suggestion
            'original' => $original,
            'root'     => __DIR__ . '/fixtures/files/test.jpg'
```

**Review by lukasbestle — COMMENTED** [review]
I really really love how clean this all became immediately. ❤️ 
I agree we should move the other filesystem-related classes as well, but we can do it in another PR.

A general comment: I feel like the `Filesystem` package should be MIT-licensed as it's not CMS specific.

**Comment by distantnative** [comment]
Great comments! Will go through them once I feel a bit more alive again

File: config/components.php
Comment (distantnative) [reply config/components.php]: @lukasbestle I thought it goes together well with `$src`. And that we use it the same way e.g. in the `App` class.

Comment (lukasbestle) [reply config/components.php]: Right, makes sense.

File: src/Toolkit/File.php
Comment (distantnative) [reply src/Toolkit/File.php]: Would mean the code would need to look something like

```php
$response = new Response('', $this->mime());

if ($send !== true) {
    return $response;
}

$result = $response->send();
return empty($result) !== true ? $result : null;
```

IMO I am not sure if this is such a good change since it differs how other parts like `Response::send` work. In general, it makes sense but we probably should change it in more places then (and on a separate PR).

**Comment by lukasbestle** [comment]
I've rebased the PR against `develop` as somehow a lot of unrelated commits ended up in here. Also I've bundled all suggestion fixes into one single commit to make the commit history less confusing.

File: src/Toolkit/File.php
Comment (lukasbestle) [reply src/Toolkit/File.php]: The `$result` is always empty as it's the body that we explicitly pass as `''` in the first line. So the code would be this:

```php
$response = new Response('', $this->mime());

if ($send !== true) {
    return $response;
}

$response->send();
```

Comment (lukasbestle) [reply src/Toolkit/File.php]: I think this is the only place where `Response` is only used to send headers. So I don't think this behavior would interfere with other uses.

But I'm open about leaving this as it is.

**Comment by distantnative** [comment]
@lukasbestle Do you know a way to tell Github "commit all these selected suggested changes with a single commit"?

**Comment by lukasbestle** [comment]
Yes, there is an "add to batch" button right next to the "commit" button.

**Comment by distantnative** [comment]
Another open discussion point: Should this be `MIT` license?
as we are moving more and more code out of `Kirby\Cms` into dedicated namespaces, I am not sure if we should all this `MIT`.

**Comment by lukasbestle** [comment]
I looked through the `Filesystem` code again and think that the classes we have in it at the moment are so CMS-specific that they can well stay proprietary.

If we were to move `F` and `Dir`, I think these should stay MIT though.
But then the new question would be: What happens with `Toolkit\File`? Merge it with the current `Filesystem\File`?

**Comment by distantnative** [comment]
> But then the new question would be: What happens with Toolkit\File? Merge it with the current Filesystem\File?

That's mainly why I said, let's do that in a separate PR. That's where the headaches start :D

I wouldn't mind having different licenses within a package (per class).

**Comment by lukasbestle** [comment]
Me neither. So I think the licenses of the code in the current PR are fine.

Now the `File::header()` discussion seems to be the only open one.

File: src/Toolkit/File.php
Comment (distantnative) [reply src/Toolkit/File.php]: @lukasbestle So is your suggestion to return `null` or nothing (void)? Your initial comment and your code differ there. I'm fine with either.

Comment (lukasbestle) [reply src/Toolkit/File.php]: Definitely `void`. My comment to "return `null`" was misunderstandable. I just meant that we shouldn't return anything in this case (= `void`), which automatically results in a `null` value.

I've changed it. :)

**Review by lukasbestle — DISMISSED** [review]

**Comment by lukasbestle** [comment]
It's really hard to track all changes across this PR, but I don't see any obvious issues. So let's roll with it. Again, it's really awesome how clean this all became. 🎉

File: src/Filesystem/Asset.php
Comment (lukasbestle) [inline src/Filesystem/Asset.php:23]: Oh, one thing I just noticed: Why was `FileModifications` removed?

Comment (distantnative) [reply src/Filesystem/Asset.php:23]: Good catch. I think reviewing all the code, I was really convinced that `FileModifications` in there was a mistake and wouldn't actually work. Now looking at it a week later, I think it does work and I was mistaken. We should add it back in then.

Comment (lukasbestle) [reply src/Filesystem/Asset.php:23]: What I think is also a bit weird is that the `Asset` class gets an `asset()` method from `IsFile`. But as that's always been the case, I think we should leave it in to avoid even more breaking changes.

Comment (distantnative) [reply src/Filesystem/Asset.php:23]: Yea, I have also moved in circles with the naming here. You quickly end up with having a `file()` method in `File` etc.

Comment (distantnative) [reply src/Filesystem/Asset.php:23]: @lukasbestle have added it again

File: src/Filesystem/Asset.php
Comment (lukasbestle) [inline src/Filesystem/Asset.php:51]: This method was basically useless before, but I wonder why it was there. Probably for compatibility, but with what? Do we need a replacement for it?

File: src/Image/Image.php
Comment (lukasbestle) [inline src/Image/Image.php]: It's not the URL but the HTML `` code. I'm thinking: Do we really need and want this breaking change?

File: tests/Filesystem/FilenameTest.php
Comment (lukasbestle) [inline tests/Filesystem/FilenameTest.php]:
```suggestion
class FilenameTest extends TestCase
```

**Review by lukasbestle — COMMENTED** [review]
Last round of comments after another detailed review. 😅
After that we can finally merge it. 🎉

File: src/Filesystem/Asset.php
Comment (bastianallgeier) [reply src/Filesystem/Asset.php:51]: If I remember correctly this is making sure that the img html can still be created correctly. For CMS files we have the alt attribute coming from the content file, but for assets, such a related content object does not exist.

Comment (lukasbestle) [reply src/Filesystem/Asset.php:51]: Right! `FileFoundation` called the `alt()` method from its `html()` method. `IsFile` doesn't have that method, instead it uses the one from `Filesystem\File` that no longer relies on `alt()` as that use-case was moved to `Image\Image`.

Removing the `alt()` method should therefore be safe IMO, but it's still a breaking-change.

File: src/Image/Image.php
Comment (distantnative) [reply src/Image/Image.php]: Changed it back to URL

File: src/Filesystem/Asset.php
Comment (distantnative) [reply src/Filesystem/Asset.php:51]: Exactly, I think it should be fine.

Comment (lukasbestle) [reply src/Filesystem/Asset.php:51]: I've added the breaking change to the first post.

**Review by lukasbestle — APPROVED** [review]

**Comment by distantnative** [comment]
@bastianallgeier I'll let you have a final look to decide if we should merge now. Lukas and I are a bit anxious cause it is a big one but at the same time we think it's ready and shouldn't contain any bigger regressions anymore.

**Comment by bastianallgeier** [comment]
It all looks very good to me. I think we still need to work on the F and Dir classes and especially on the CMS file stuff next to make it all fit together, but we are already quite deep in heavy changes that need careful launch preparations, once this is all finished. We should write down very clear instructions for everyone who might run into issues and we need a long and active test phase for this release.

**Review by bastianallgeier — APPROVED** [review]

**Comment by distantnative** [comment]
@bastianallgeier moving F and Dir is quite straight forward. Why we have shied away from it so far is more `Toolkit\File` - that's one to really think about

**Comment by lukasbestle** [comment]
I'm pretty confident we have collected all breaking changes that are relevant, so IMO we are on a good path.

**Pull request merged**
